x/programs/rust/wasmlanche_sdk/src/simulator.rs

```diff
@@ -0,0 +1,326 @@
+    #[serde(untagged)]
+    Key(Key),
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+#[serde(rename_all = "lowercase")]
+pub enum Key {
+    Ed25519,
+    Secp256r1,
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+pub struct Param<'a> {
+    /// The optional name of the parameter. This is only used for readability.
+    name: &'a str,
+    #[serde(rename = "type")]
+    /// The type of the parameter.
+    param_type: ParamType,
+    /// The value of the parameter.
+    value: &'a str,
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+pub struct Require {
+    /// If defined the result of the step must match this assertion.
+    result: ResultAssertion,
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+pub enum Operator {
+    #[serde(rename = "==")]
+    NumericEq,
+    #[serde(rename = "!=")]
+    NumericNe,
+    #[serde(rename = ">")]
+    NumericGt,
+    #[serde(rename = "<")]
+    NumericLt,
+    #[serde(rename = ">=")]
+    NumericGe,
+    #[serde(rename = "<=")]
+    NumericLe,
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+pub struct ResultAssertion {
+    /// The operator to use for the assertion.
+    operator: Operator,
+    /// The value to compare against.
+    value: String,
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+pub struct Plan<'a> {
+    /// The name of the plan.
+    name: &'a str,
+    /// A description of the plan.
+    description: &'a str,
+    /// The key of the caller used in each step of the plan.
+    caller_key: &'a str,
+    /// The steps to perform in the plan.
+    steps: Vec<Step<'a>>,
+}
+
+pub struct Client {
+    /// Path to the simulator binary
+    path: String,
+}
+
+impl Client {
+    #[must_use]
+    pub fn new(path: String) -> Self {
+        Self { path }
+    }
+
+    /// Runs a `Plan` against the simulator and returns the result.
+    /// # Errors
+    ///
+    /// Returns an error if the if serialization or plan fails.
+    pub fn run<T>(&self, plan: &Plan) -> Result<T, Box<dyn Error>>
+    where
+        T: serde::de::DeserializeOwned + serde::Serialize,
+    {
+        call_run_stdin(&self.path, plan)
+    }
+
+    /// Performs a `ReadOnly` step against the simulator and returns the result.
+    /// # Errors
+    ///
+    /// Returns an error if the if serialization or plan fails.
+    pub fn read_only<T>(&self, data: Step, key: &str) -> Result<T, Box<dyn Error>>
+    where
+        T: serde::de::DeserializeOwned + serde::Serialize,
+    {
+        let plan = &Plan {
+            name: "view",
+            description: "single view request",
+            caller_key: key,
+            steps: vec![data],
+        };
+
+        call_run_stdin(&self.path, plan)
+    }
+
+    /// Performs a single `Execute` step against the simulator and returns the result.
+    /// # Errors
+    ///
+    /// Returns an error if the if serialization or plan fails.
+    pub fn execute<T>(&self, data: Step, key: &str) -> Result<T, Box<dyn Error>>
```

Comment:
what is the point of explicitly noting `execute` vs `read_only`?

Reply:
the expectation is that one returns a tx the other returns the result of a function.

Reply:
I think we will want the SDK to be able to better describe this. Some functions are read_only thus can be run against a local nodes as a read only call. But something that explicitly could change state would be part of a transaction. That's my thinking at least.